File: rainbow-lib/src/main/java/com/juankysoriano/rainbow/core/matrix/RMatrix.java
/* -*- mode: java; c-basic-offset: 2; indent-tabs-mode: nil -*- */

/*
  Part of the Processing project - http://processing.org

  Copyright (c) 2005-10 Ben Fry and Casey Reas

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License version 2.1 as published by the Free Software Foundation.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General
  Public License along with this library; if not, write to the
  Free Software Foundation, Inc., 59 Temple Place, Suite 330,
  Boston, MA  02111-1307  USA
*/

package com.juankysoriano.rainbow.core.matrix;

public interface RMatrix {

    void reset();

    /**
     * Returns a copy of this PMatrix.
     */
    RMatrix get();

    /**
     * Copies the matrix contents into a float array.
     * If target is null (or not the correct size), a new array will be created.
     */
    float[] get(float[] target);

    void set(RMatrix src);

    void set(float[] source);

    void set(float m00, float m01, float m02,
             float m10, float m11, float m12);

    void set(float m00, float m01, float m02, float m03,
             float m10, float m11, float m12, float m13,
             float m20, float m21, float m22, float m23,
             float m30, float m31, float m32, float m33);

    void translate(float tx, float ty);

    void translate(float tx, float ty, float tz);

    void rotate(float angle);

    void rotateX(float angle);

    void rotateY(float angle);

    void rotateZ(float angle);

    void rotate(float angle, float v0, float v1, float v2);

    void scale(float s);

    void scale(float sx, float sy);

    void scale(float x, float y, float z);

    void shearX(float angle);

    void shearY(float angle);

    /**
     * Multiply this matrix by another.
     */
    void apply(RMatrix source);

    void apply(RMatrix2D source);

    void apply(RMatrix3D source);

    void apply(float n00, float n01, float n02,
               float n10, float n11, float n12);

    void apply(float n00, float n01, float n02, float n03,
               float n10, float n11, float n12, float n13,
               float n20, float n21, float n22, float n23,
               float n30, float n31, float n32, float n33);

    /**
     * Apply another matrix to the left of this one.
     */
    void preApply(RMatrix2D left);

    void preApply(RMatrix3D left);

    void preApply(float n00, float n01, float n02,
                  float n10, float n11, float n12);

    void preApply(float n00, float n01, float n02, float n03,
                  float n10, float n11, float n12, float n13,
                  float n20, float n21, float n22, float n23,
                  float n30, float n31, float n32, float n33);

    /**
     * Multiply a PVector by this matrix.
     */
    RVector mult(RVector source, RVector target);

    /**
     * Multiply a multi-element vector against this matrix.
     */
    float[] mult(float[] source, float[] target);

    /**
     * Transpose this matrix.
     */
    void transpose();

    /**
     * Invert this matrix.
     *
     * @return true if successful
     */
    boolean invert();

    /**
     * @return the determinant of the matrix
     */
    float determinant();
}

File: rainbow-lib/src/main/java/com/juankysoriano/rainbow/core/math/RMatrix.kt
/* -*- mode: java; c-basic-offset: 2; indent-tabs-mode: nil -*- */

/*
  Part of the Processing project - http://processing.org

  Copyright (c) 2005-10 Ben Fry and Casey Reas

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License version 2.1 as published by the Free Software Foundation.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General
  Public License along with this library; if not, write to the
  Free Software Foundation, Inc., 59 Temple Place, Suite 330,
  Boston, MA  02111-1307  USA
*/

package com.juankysoriano.rainbow.core.math

interface RMatrix {

    fun reset()

    /**
     * Returns a copy of this PMatrix.
     */
    fun get(): RMatrix

    /**
     * Copies the matrix contents into a float array.
     * If target is null (or not the correct size), a new array will be created.
     */
    fun toFloatArray(): FloatArray

    fun set(src: RMatrix)

    fun set(source: FloatArray)

    fun set(m00: Float, m01: Float, m02: Float,
            m10: Float, m11: Float, m12: Float)

    fun set(m00: Float, m01: Float, m02: Float, m03: Float,
            m10: Float, m11: Float, m12: Float, m13: Float,
            m20: Float, m21: Float, m22: Float, m23: Float,
            m30: Float, m31: Float, m32: Float, m33: Float)

    fun translate(tx: Float, ty: Float)

    fun translate(tx: Float, ty: Float, tz: Float)

    fun rotate(angle: Float)

    fun rotateX(angle: Float)

    fun rotateY(angle: Float)

    fun rotateZ(angle: Float)

    fun rotate(angle: Float, v0: Float, v1: Float, v2: Float)

    fun scale(s: Float)

    fun scale(sx: Float, sy: Float)

    fun scale(x: Float, y: Float, z: Float)

    fun shearX(angle: Float)

    fun shearY(angle: Float)

    /**
     * Multiply this matrix by another.
     */
    fun apply(source: RMatrix)

    fun apply(source: RMatrix2D)

    fun apply(source: RMatrix3D)

    fun apply(n00: Float, n01: Float, n02: Float,
              n10: Float, n11: Float, n12: Float)

    fun apply(n00: Float, n01: Float, n02: Float, n03: Float,
              n10: Float, n11: Float, n12: Float, n13: Float,
              n20: Float, n21: Float, n22: Float, n23: Float,
              n30: Float, n31: Float, n32: Float, n33: Float)

    /**
     * Apply another matrix to the left of this one.
     */
    fun preApply(left: RMatrix2D)

    fun preApply(left: RMatrix3D)

    fun preApply(n00: Float, n01: Float, n02: Float,
                 n10: Float, n11: Float, n12: Float)

    fun preApply(n00: Float, n01: Float, n02: Float, n03: Float,
                 n10: Float, n11: Float, n12: Float, n13: Float,
                 n20: Float, n21: Float, n22: Float, n23: Float,
                 n30: Float, n31: Float, n32: Float, n33: Float)

    /**
     * Multiply a PVector by this matrix.
     */
    fun mult(source: RVector, target: RVector): RVector

    /**
     * Multiply a multi-element vector against this matrix.
     */
    fun mult(source: FloatArray, target: FloatArray): FloatArray

    /**
     * Transpose this matrix.
     */
    fun transpose()

    /**
     * Invert this matrix.
     *
     * @return true if successful
     */
    fun invert(): Boolean

    /**
     * @return the determinant of the matrix
     */
    fun determinant(): Float
}
